vic/drivers/shared_image/src/vic_init_output.c

```diff
@@ -24,270 +24,303 @@
+    extern option_struct      options;
+    extern domain_struct      local_domain;
+    extern filep_struct       filep;
+    extern MPI_Comm           MPI_COMM_VIC;
+    extern int                mpi_rank;
+    extern all_vars_struct   *all_vars;
+    extern atmos_data_struct *atmos;
+    extern lake_con_struct    lake_con;
+    extern save_data_struct  *save_data;
+    extern soil_con_struct   *soil_con;
+    extern veg_con_struct   **veg_con;
+    extern veg_lib_struct   **veg_lib;
+    extern double          ***out_data;
+    extern stream_struct     *output_streams;
+    extern nc_file_struct    *nc_hist_files;
+    extern MPI_Datatype       mpi_alarm_struct_type;
+
+    size_t                    streamnum;
+    int                       status;
+    size_t                    i;
 
     // initialize the output data structures
+    set_output_met_data_info();
+
+    alloc_out_data(local_domain.ncells_active, &out_data);
+
     for (i = 0; i < local_domain.ncells_active; i++) {
-        put_data(&(all_vars[i]), &(atmos[i]), &(soil_con[i]), veg_con[i],
-                 veg_lib[i], &lake_con, out_data[i], &(save_data[i]),
-                 -global_param.nrecs);
+        initialize_save_data(&(all_vars[i]), &(atmos[i]), &(soil_con[i]),
+                             veg_con[i], veg_lib[i], &lake_con, out_data[i],
+                             &(save_data[i]));
     }
-
     if (mpi_rank == 0) {
         // determine which variables will be written to the history file
-        parse_output_info(filep.globalparam, out_data);
+        parse_output_info(filep.globalparam, &output_streams,
+                          local_domain.ncells_active, dmy_current);
+    }
 
-        // open the netcdf history file
-        initialize_history_file(&nc_hist_file);
+    status = MPI_Bcast(&(options.Noutstreams), 1, MPI_AINT, 0, MPI_COMM_VIC);
+    if (status != MPI_SUCCESS) {
+        log_err("MPI error %d\n", status);
     }
 
-    // broadcast which variables to write.
-    for (i = 0; i < N_OUTVAR_TYPES; i++) {
-        status = MPI_Bcast(&out_data[0][i].write, 1, MPI_C_BOOL,
-                           0, MPI_COMM_VIC);
-        if (status != MPI_SUCCESS) {
-            log_err("MPI error in vic_init_output(): %d\n", status);
-        }
+    // allocate netcdf history files array
+    nc_hist_files = calloc(options.Noutstreams, sizeof(*nc_hist_files));
+    if (nc_hist_files == NULL) {
+        log_err("Memory allocation error");
     }
 
-    // broadcast history file info. Only the master process will write to it,
-    // but the slave processes need some of the information to initialize as
-    // well (particularly which variables to write and dimension sizes)
-    status = MPI_Bcast(&nc_hist_file, 1, mpi_nc_file_struct_type,
-                       0, MPI_COMM_VIC);
-    if (status != MPI_SUCCESS) {
-        log_err("MPI error in vic_init_output(): %d\n", status);
+    // Allocate memory for array of streams
+    if (mpi_rank != 0) {
```

Comment:
Why do you not need to allocate this for `mpi_rank == 0`? I cannot think of a case where you would need it on the nodes but not for `mpi_rank ==  0`. If you already did it for `mpi_rank == 0` elsewhere then you should do all of them at the same time (i.e. no check on `mpi_rank` should be necessary here.